extensions/default/src/Components/DataSourceConfigurationModalComponent.tsx

```diff
@@ -0,0 +1,209 @@
+   * full/existing configuration for a data source is to be shown, the current
+   * selected item is the second to last in the `selectedItems` list.
+   */
+  const currentSelectedItemIndex = showFullConfig
+    ? selectedItems.length - 2
+    : selectedItems.length - 1;
+
+  useEffect(() => {
+    let shouldUpdate = true;
+
+    setErrorMessage(null);
+
+    // Clear out the former/old list while we fetch the next sub item list.
+    setItemList(null);
+
+    if (selectedItems.length === 0) {
+      configurationAPI
+        .initialize()
+        .then(items => {
+          if (shouldUpdate) {
+            setItemList(items);
+          }
+        })
+        .catch(error => setErrorMessage(error.message));
+    } else if (!showFullConfig && selectedItems.length === itemLabels.length) {
+      // The last item to configure the data source (path) has been selected.
+      configurationAPI.setCurrentItem(selectedItems[selectedItems.length - 1]);
+      // We can hide the modal dialog now.
+      onHide();
+    } else {
+      configurationAPI
+        .setCurrentItem(selectedItems[currentSelectedItemIndex])
+        .then(items => {
+          if (shouldUpdate) {
+            setItemList(items);
+          }
+        })
+        .catch(error => setErrorMessage(error.message));
+    }
+
+    return () => {
+      shouldUpdate = false;
+    };
+  }, [
+    selectedItems,
+    configurationAPI,
+    onHide,
+    itemLabels,
+    showFullConfig,
+    currentSelectedItemIndex,
+  ]);
+
+  const getErrorComponent = (): ReactElement => {
+    return (
+      <div className="flex flex-col gap-4 min-h-[1px] grow">
+        <div className="text-primary-light text-[20px]">
+          {t(`Error fetching ${itemLabels[selectedItems.length]} list`)}
+        </div>
+        <div className="bg-black text-[14px] grow p-4">{errorMessage}</div>
+      </div>
+    );
+  };
+
+  const getSelectedItemsComponent = (): ReactElement => {
+    const getCursorClasses = itemIndex =>
+      itemIndex !== itemLabels.length - 1 && itemIndex < selectedItems.length
+        ? 'cursor-pointer'
+        : 'cursor-auto';
+
+    const getBackgroundClasses = itemIndex =>
+      itemIndex < selectedItems.length
+        ? classNames(
+            'bg-black/[.4]',
+            itemIndex !== itemLabels.length - 1
+              ? 'hover:bg-transparent active:bg-secondary-dark'
+              : ''
+          )
+        : 'bg-transparent';
+
+    const getBorderClasses = itemIndex =>
+      itemIndex === currentSelectedItemIndex + 1
+        ? classNames('border-2', 'border-solid', 'border-primary-light')
+        : itemIndex < selectedItems.length
+        ? 'border border-solid border-primary-active hover:border-primary-light active:border-white'
+        : 'border border-dashed border-secondary-light';
+
+    const getTextClasses = itemIndex =>
+      itemIndex <= selectedItems.length
+        ? 'text-primary-light'
+        : 'text-primary-active';
+
```

Comment:
I still think these should live outside of this component as they are just adding noise to reading

Reply:
I think perhaps moving them outside of the `getSelectedItemsComponent` makes sense, but notice that they rely on various states (i.e. `currentSeletedItemIndex` and `itemLabels.length`) of the `DataSourceConfigurationModalComponent` that I think would make it awkward moving them outside of that. I will update, push and we can go from there.